stl/inc/__msvc_system_error_abi.hpp

```diff
@@ -0,0 +1,43 @@
+// __msvc_system_error_abi.h internal header (core)
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef __MSVC_SYSTEM_ERROR_ABI_HPP
+#define __MSVC_SYSTEM_ERROR_ABI_HPP
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+#include <stddef.h>
+

```

Comment:
In this case I believe marking the header `(core)` is sufficient to convey that it may be included in the import libraries. (After all, the point of `(core)` headers is that they may be included in any context without restriction)

Of course `<xfacet>` cannot be made `(core)` because it depends on the "DLL-ness" of "P".

Reply:
Having lived through the Scary Bad Times*, I strongly believe that we need the Ominous Comment as a reminder to watch our dependencies even more closely than usual. I don't yet think that we pay sufficient attention to `(core)`. The danger is that because this is an internal-only header, dragged in by `<system_error>` (thoroughly non-core), losing its core-ness would not be an immediate issue, the way it would be for `<type_traits>`. For things that aren't enforced automatically, we need good comment discipline, so contributors can clearly see what changes are valid, and we can easily notice when such restrictions aren't being followed.

If we changed our convention for core headers, e.g. introducing a brief comment explaining what they are, and what dependencies can be included, then I would no longer think that we need a separate ominous comment here.

(* It was extremely bad. The ODR violation went undetected for a long time, and had symptoms only in the debugger visualizations.)

Reply:
>this is an internal-only header, dragged in by <system_error> (thoroughly non-core)

Right, but all the `xxx_abi.hpp` headers are core; that's why they're here. Note that the status quo is that `<xfilesystem_abi.h>` is in the import lib, and it includes all manner of things, and we don't mark all of them with the omnibus comment.

>If we changed our convention for core headers, e.g. introducing a brief comment explaining what they are, and what dependencies can be included, then I would no longer think that we need a separate ominous comment here.

Could we fold the "omnibus comment" into https://github.com/microsoft/STL/wiki/What's-the-difference-between-(core)-and-non-(core)-headers%3F and then link to it from all the core headers? I worry that the omnibus comment is long enough to be "skimmed over" boilerplate and also slightly inaccurate -- the length of the file has no bearing on what is correct to go in there. The only thing that matters is not depending on our configuration macros like `_ITERATOR_DEBUG_LEVEL`.

As an aside, do you remember why `pragma detect_mismatch` didn't detect the Scary Bad Times condition?

Reply:
I'm not sure why `#pragma detect_mismatch` didn't detect the problem, but I think the reason was that we build the STL with the suppression macro.

(Note: ominous, not omnibus.)

I think during the review of Filesystem I mentioned that I was somewhat uncomfortable with the sheer size of stuff that was being injected into the import lib, but it's true, only things that are macro-sensitive are especially problematic.

I disagree that long comments are skimmable. We really refrain from long comments in the STL, so their presence indicates that something unusual is going on. Perhaps it should be more SHOUTY, though. I don't think that concealing it behind a link would be an improvement.

Reply:
>I'm not sure why #pragma detect_mismatch didn't detect the problem, but I think the reason was that we build the STL with the suppression macro.

That's probably why.

I filed an issue about this ( https://github.com/microsoft/STL/issues/483 ) ; will add the comment to the .cpp only.

Reply:
Actually the .cpp has this comment already.